### DISCUSSION TITLE
Using tree-sitter to assist LSP completion

### DISCUSSION THREAD

**Discussion by Ce11an** [tree-sitter/tree-sitter#3346] [Q&A] [answered]
Hey  :wave: 

I hope it's ok to ask questions here! 

I am working on a creating a tree-sitter for SurrealQL the query language of SurrealDB. I have started the [grammar](https://github.com/Ce11an/tree-sitter-surrealql/blob/main/grammar.js). It is very much a work in progress. While working on the grammar, I wanted to see if I could use my newly created tree-sitter as a parser for an LSP I am building. I am trying to implement completion and failing miserably (this is my first time with tree-sitter and LSPs :laughing: ). Does anyone have any advice? I have some examples of what I am trying to achieve:

```sql
"X"  is a place holder for where the cursor will be:

1. SELECT X

Completion list would be ["VALUE"]

2. SELECT amount FROM payments X

Completion list should be ["ORDER BY", "WHERE", "SPLIT", "GROUP BY"]

3. SELECT amount FROM payments:cellan X

Completion list should be ["ORDER BY", "WHERE", "SPLIT", "GROUP BY"]
```
Essentially, I want to get the closest previous node to the cursor. I can map the node to the values I want to populate for completion. However, when I parse the code the node kind for value X "source_file". Do I need to walk the cursor, should I query the tree, or  should I take a completely different approach? I am not sure if it is the way my grammar is set up or I am going about this the wrong way.

 Any advice would be appreciated! 🙏🏻

**Comment by WillLillis** [comment, marked as answer]
I added some rudimentary completion capabilities to [asm-lsp](https://github.com/bergercookie/asm-lsp) using tree-sitter's rust bindings. The basic building blocks for this were:

- Maintain an in-memory copy of the source file being edited. This copy can be updated in step with edits made to the source file in [two ways](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_synchronization):
    - Incremental -- The LSP client sends edits over that you then apply to your copy. If you're writing the LSP in Rust the [lsp-textdocument](https://crates.io/crates/lsp-textdocument) crate makes this fairly straightforward
    - Full -- The client sends a full copy of the document each time it's edited. This is much simpler but also comes at a performance cost, especially as the documents grow larger.
- Alongside the in-memory copy of the source file, maintain a `tree-sitter` tree. If you're going with incremental document synchronization, you can also take advantage of tree-sitter's incremental parsing by applying edits to the tree and reparsing, rather than reparsing the entire tree each time an edit is made. 
- Each time a [Completion request](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_completion) is issued to the LSP server, the program would then issue a query checking for nodes under/nearby the current cursor position that the server is able to provide autocomplete suggestions for (in the case of asm-lsp, this is currently just registers and instructions). Depending on how you write your grammar, one way you could approach this is to make all of the `X`'s from your example as children of the parent nodes (e.g. `SELECT`, `FROM`, etc.). If your cursor overlaps with a child field's location, you can offer suggestions based on the node type or textual content of the corresponding parent. 

To give some more concrete info, in asm-lsp: 

- The completion logic is all computed [here](https://github.com/bergercookie/asm-lsp/blob/0a274567e908c1ae293300c177c0b62ccd454720/src/lsp.rs#L365-L460)
- The update logic is handled [here](https://github.com/bergercookie/asm-lsp/blob/0a274567e908c1ae293300c177c0b62ccd454720/src/bin/main.rs#L408-L435)

Hope this is helpful :)

**Reply by Ce11an** [reply]
> suggestions

Wow, that's a detailed response - thank you! I've been trying to walk the cursor find the relevant matches, but I'll try to use a query now 🙌🏻 

In your opinion, would an LSP for a SQL like language be primarily trigger based, i.e. if certain keywords are written followed by a whitespace? I'll definitely go through your examples for inspiration ✨

**Reply by WillLillis** [reply]
>  I've been trying to walk the cursor find the relevant matches, but I'll try to use a query now 🙌🏻

Yeah, when I started I initially tried manually traversing the tree too, but found the query API to be *significantly* easier. I would also highly recommend the cli tool's playground feature for testing out queries for your grammar.

> In your opinion, would an LSP for a SQL like language be primarily trigger based, i.e. if certain keywords are written followed by a whitespace?

When I was working on completion, I worried less about how the requests were triggered and more of just shaping the program around how it would respond to [Requests](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#requestMessage) and [Notifications](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#notificationMessage) from the LSP spec. It was a lot easier to rely on tree-sitter to determine when/how to provide completions based on the source file at a syntactic level, rather than a textual (with the small exception of [trigger characters](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionTriggerKind)). There's definitely a lot of ways you could do this, but I would recommend starting by trying to match against certain node patterns with a tree-sitter query and then offering completions from that point. For a quick example pulling from your grammar, a simple query could be 

```
(source_file
  (select_statement
    (select_clause)
    (from_clause) . (_)?))
```

If the query matches and the cursor column position is past the end of the `(from_clause)` node, you could return a completion list for all of your optional clauses (e.g. `WITH`, `WHERE`, etc.).

**Reply by Ce11an** [reply]
Thanks for the advice 😮 I managed to get the playground working via the CLI, should have done this from the start - so nice! I will play around with this query more when implementing it into the LSP 💯 
![Screenshot 2024-05-08 at 20 38 35](https://github.com/tree-sitter/tree-sitter/assets/60790416/a38d46da-9e56-4555-810e-39520265fafd)

**Reply by Ce11an** [reply]
Hey @WillLillis! 👋🏻

I was hoping to get some more insights from you 🙏🏻 I've been tinkering with this more, and the good news is I've successfully integrated my LSP with Neovim using tower-lsp and lsp-textdocument 🥳 However, I'm encountering an issue with matching my queries to the cursor position. For instance, when I write:

```sql
SELECT * FROM table <cursor-is-here>
```

Since all the keywords I want to offer completions for are [optional nodes](https://github.com/Ce11an/tree-sitter-surrealql/blob/ad1638d211710ca662d1c83b3e333accc11d8a9d/grammar.js#L76), my query doesn't capture the whitespace where my cursor is, making it unable to offer completions. I can match a cursor that is after a capture, but I need to handle multiple queries/matches. Currently, I have a mapping between the capture name and the completions I can offer. Do you have any suggestions for what I could do in this situation?

I'm happy to provide more context if needed. Thanks again for your help!

**Reply by WillLillis** [reply]
> Hey @WillLillis! 👋🏻
> 
> I was hoping to get some more insights from you 🙏🏻 I've been tinkering with this more, and the good news is I've successfully integrated my LSP with Neovim using tower-lsp and lsp-textdocument 🥳 

Hey! That's great to hear! 

> However, I'm encountering an issue with matching my queries to the cursor position. For instance, when I write:
> 
> ```sql
> SELECT * FROM table <cursor-is-here>
> ```
> 
> Since all the keywords I want to offer completions for are [optional nodes](https://github.com/Ce11an/tree-sitter-surrealql/blob/ad1638d211710ca662d1c83b3e333accc11d8a9d/grammar.js#L76), my query doesn't capture the whitespace where my cursor is, making it unable to offer completions. 

I'm not sure I fully understand. The way I'd approach this problem would be a query somewhere along the lines of 

```scm
(from_clause (target) @target .)
```

With the capture `@target`, you would then have the range of text that `table` occupies. Also note the trailing anchor, `.`, that ensures the query only matches if the `(target)` node is the last child of the `(from_clause)` nodes. If your cursor position is past that range (but on the same line), you could offer completions for the keywords of all of the optional nodes in your grammar. A separate query for each of the optional nodes could then be matched against later on to offer subsequent completions for the optional nodes' sub parts (e.g. `by` for the `limit_clause`).

> I can match a cursor that is after a capture, but I need to handle multiple queries/matches. Currently, I have a mapping between the capture name and the completions I can offer. Do you have any suggestions for what I could do in this situation?

I'm not sure if this describes the solution I gave above or is something different. Could you give a few examples? If you're having trouble telling which capture matched in a given query, an easy option would be to split the query up into multiple separate queries (to be issued one after the other, short circuiting once one of them matches) where you can handle the behavior of each query separately a little easier.

I'm happy to take a closer look and dig into this further with some more context/ examples :). If you're comfortable sharing the LSP's code I'd be happy taking a look at that as well!

**Reply by Ce11an** [reply]
Hey @WillLillis - thanks for the speedy response ⚡ 

I just made the [repo](https://github.com/Ce11an/surrealql-lsp) public if you want to take a look 👁️. Essentially, I have a few current issues due to my lack of tree-sitter query knowledge and the typical ways of building LSPs (thanks for your patience 🙏🏻):

- The repo contains a simple example of completion. I check for a match, and if the cursor is after the match, I provide a completion. This works for suggesting `WHERE` and `SPLIT`. However, I am unsure how to implement another query to match `SELECT` and return `VALUE`, especially if there are multiple matches on the same line.
- Typically, SQL-like languages are written on multiple lines. What do you think is the best way to handle this with matches and cursor position?

I hope this clarification helps! Again, thanks for your support. I’m happy to move this from a discussion if it’s more appropriate! 💯 

![Screenshot 2024-07-01 at 20 42 39](https://github.com/tree-sitter/tree-sitter/assets/60790416/63be0d7c-4c70-4ec9-b8ae-5e46e244ac62)

**Reply by WillLillis** [reply]
> Hey @WillLillis - thanks for the speedy response ⚡
> 
> I just made the [repo](https://github.com/Ce11an/surrealql-lsp) public if you want to take a look 👁️.  Essentially, I have a few current issues due to my lack of tree-sitter query knowledge and the typical ways of building LSPs (thanks for your patience 🙏🏻):

Great! And no worries, I really enjoy this stuff!
 
> * The repo contains a simple example of completion. I check for a match, and if the cursor is after the match, I provide a completion. This works for suggesting `WHERE` and `SPLIT`. However, I am unsure how to implement another query to match `SELECT` and return `VALUE`, especially if there are multiple matches on the same line.

I think I see what you mean, this should be pretty doable. I'll set aside some time to play with this and (hopefully) figure it out soon. Could you send a couple examples with the existing text, cursor position, and the autocomplete suggestions you're expecting to see?

> * Typically, SQL-like languages are written on multiple lines. What do you think is the best way to handle this with matches and cursor position?

Oh yeah my bad there. Checking if the cursor falls within a match's range should still be relatively straightforward here, you just can't restrict the query to only look at the cursor's current line.

> I hope this clarification helps! Again, thanks for your support. I’m happy to move this from a discussion if it’s more appropriate! 💯

Happy to help! I think it would probably be better to move the discussion over to your repo, just so we aren't pinging the other tree-sitter folks. I'll be able to spend some time looking into this soon :)

**Reply by Ce11an** [reply]
A small update - I have bullet point one working! Essentially, I loop through all the captures where the cursor matches and keep track of the last one that matched, providing completion only for that. It's not perfect, but it's a start! 🚀
